### ISSUE TITLE
Automatic build

### ISSUE THREAD

**Issue by mildred** [purpleidea/mgmt#113]
Would it be possible to publish automated builds generated by travis on GitHub Releases ?

https://docs.travis-ci.com/user/deployment/releases/

This would allow us to download the binary for easy deploy on servers.

**Comment by purpleidea** [comment]
@mildred I'm happy to start doing this, although I'd rather build the binaries off of travis and push them somewhere. We do have https://copr.fedorainfracloud.org/coprs/purpleidea/mgmt which needs updating since .spec file is broken. Would that work for you??

**Comment by mildred** [comment]
I specifically need a static binary (and I don't need libvirt), because I deploy mgmt on CoreOS. I'm afraid rpms won't work for that. What I had in mind was a binary distribution that can be independent of the distribution, and can be downloaded with a simple wget/curl.

We could provide:

- static binary without libvirt or augeas
- zip file containing dynamic binary with its library dependencies

I can prepare a build script to generate these binary packages if you want.

**Comment by purpleidea** [comment]
@mildred We have build tags in git master now: https://github.com/purpleidea/mgmt/blob/master/docs/documentation.md#compilation-options

**Comment by mildred** [comment]
Yes, I know and used them to ger mgmt on CoreOS.

**Comment by purpleidea** [comment]
@mildred Okay, so how about this... In the project `Makefile`, I have some targets to build rpm's, sign them, and push them over ssh to a server. If you patch the Makefile to add a new binary only build target and to push to that same server, I'll start running it on release. Does that work? Should be mostly a copy+paste job if you look at what's there and are careful.

LMK what you think!

**Comment by roidelapluie** [comment]
This:

https://github.com/prometheus/prometheus/releases/tag/v1.5.2

**Comment by purpleidea** [comment]
@roidelapluie How do I post binaries there?

**Comment by roidelapluie** [comment]
@purpleidea from travis ..

**Comment by purpleidea** [comment]
Looks like this is a GH feature: https://github.com/blog/1547-release-your-software

**Comment by purpleidea** [comment]
I'm happy to add this to the [tag.sh](https://github.com/purpleidea/mgmt/blob/master/tag.sh) script or the [Makefile](https://github.com/purpleidea/mgmt/blob/master/Makefile], but I don't want to use travis to make the builds unless they're [reproducible](https://github.com/purpleidea/mgmt/blob/master/test/test-reproducible.sh) -- although proper reproducible.

**Comment by roidelapluie** [comment]
yes but you want to automate that with travis, not just drop random
binaries

https://docs.travis-ci.com/user/deployment/releases/

**Comment by purpleidea** [comment]
@roidelapluie To make a release, I run tag.sh so I consider that automated enough for me... I am attempting to avoid an untrusted builder that has my name on the binary.

**Comment by roidelapluie** [comment]
" untrusted builder " .. you are the untrusted builder,  not travis :)

**Comment by purpleidea** [comment]
> " untrusted builder " .. you are the untrusted builder, not travis :)

@roidelapluie That's true! Except my "purpleidea" reputation is on the line... Point taken though. Let me look at the docs some more for this.

**Comment by roidelapluie** [comment]
that is why you want to move to a mgmtconfig org ?

**Comment by purpleidea** [comment]
Same difference

**Comment by purpleidea** [comment]
FYI: https://tests.reproducible-builds.org/debian/issues/unstable/golang_compiler_captures_build_path_in_binary_issue.html

**Comment by DanielHouston** [comment]
CI is my Jam - I see no reason why we can't use test-reproducable.sh as a gate to autonomous releases. Can I get view access on the travis jobs?

**Comment by purpleidea** [comment]
@DanielHouston You should not be blocked from seeing anything in CI.

**Comment by DanielHouston** [comment]
Regarding verification: The simpliest solution is using ol' fashioned public key verification. 

You own foo.com:

Upon release you run a generate-hash.sh from your developer machine; which takes your private key and to produce a hash of the latest the mgmt executable.

encrypted hash of mgmt executable pushed to foo.com/mgmt/releases/${release.x.x.x}/hash

new release is tagged and pushed to release branch

Travis build monitoring release branch builds and tests mgmt; finally pulling down foo.com/mgmt/releases/${release.x.x.x}/hash.

Travis decrypts using your public key and compares hashes.
On success, build is a green light for release; shared on Github under releases.
On failure build is aborted and no release is shared.

This gates releasing on you updating the hashes.

if you wanted to ensure the latest version being offered on GH is still safe and verified you could always have a cron or similar on a secure box which repeats the build -> verify hashes steps against the latest mgmt executable on GH and alerts you if they no longer match.
